### DISCUSSION TITLE
Proposal for non-cooperative abortion of code execution

### DISCUSSION THREAD

**Discussion by AntonLapounov** [dotnet/runtime#66480] [Ideas]
There is a number of execution environments for .NET code that may benefit from having an option to perform non-cooperative abortion of user code, e.g., C# and F# interactive compilers (`csi` and `fsi`) or .NET Interactive Notebooks (see dotnet/runtime#41291).  While the legacy Framework provided the `Thread.Abort` method, it has never been supported by .NET Core, because aborting code execution at a random point is inherently unsafe and may leave the process in an inconsistent or a corrupted state.  Thus, the only available option to interrupt stuck code execution is to terminate the process and lose its state, which is undesirable for interactive scenarios.

Below is the [proposed](https://github.com/dotnet/runtime/issues/41291#issuecomment-679617031) `System.Runtime.CompilerServices.ControlledExecution` API to allow controlled execution of given code.  The `Run` method starts execution of the delegate and `TryAbort` method may be used to attempt aborting its execution by throwing an asynchronous exception on that thread.  Aborting may be successful only when the code is executing managed code, e.g., running a CPU-intensive loop, on the original thread.  This API would not help cases where the thread is stuck in unmanaged code or waiting for some other thread to finish its work.

```C#
public class ControlledExecution
{
    public ControlledExecution(Action action);

    // Runs the specified action.
    // Throws an InvalidOperationException if the action is already running.
    // Throws a ThreadAbortException if execution was aborted.
    public void Run();

    // Attempts to abort the execution of the action by injecting an asynchronous exception.
    // Returns true on success or false otherwise.
    // Throws an InvalidOperationException if the action is not running.
    public bool TryAbort(TimeSpan timeout);
}
```
The implementation will generally use the same mechanism as the one utilized by the debugger to abort threads doing function evaluation ([`Thread::UserAbort`](https://github.com/dotnet/runtime/blob/c8eeba753183d07679b7d2e65cc12d39c948eb4a/src/coreclr/vm/threadsuspend.cpp#L1175)).  That includes marking the thread for abortion, suspending it, checking whether it is running managed code and safe to redirect, redirecting the thread to the [`ThrowControlForThread`](https://github.com/dotnet/runtime/blob/c8eeba753183d07679b7d2e65cc12d39c948eb4a/src/coreclr/vm/threadsuspend.cpp#L3880) runtime helper, walking the stack to verify the tread is in a safe spot, and raising the exception for thread abort.

A possible option to make the implementation simpler is to rely only on GC polls / polls for thread abort instead of redirecting the thread on Windows, similarly to what is done on Unixes.

**Comment by stephentoub** [comment]
We've deleted a whole bunch of code in .NET Core that was there to help try to improve reliability in the face of thread aborts, e.g. we might now have:
```C#
_spinLock.Enter();
_i++;
_j++;
_spinLock.Exit();
```
instead of:
```C#
bool entered = false;
try
{
    _spinLock.Enter(ref entered);
    _i++;
    _j++;
}
finally
{
    if (entered)
    {
        _spinLock.Exit();
    }
}
```
New code has also been written assuming no thread aborts.  Is the proposal that all such code needs to come back?  Or is the idea that if a ThreadAbortException is thrown out of Run, the caller should promptly tear down the process?

**Reply by AntonLapounov** [reply]
This class will not provide any consistency guarantees and it should not be used in product environments. It may still be useful in development scenarios like interactive evaluation.

**Reply by Clockwork-Muse** [reply]
... if it's in development scenarios, what's wrong with manually killing the process (which would also take care of unmanaged code)?

**Reply by AntonLapounov** [reply]
Killing the process and losing its state is too harsh price for making a typo that led to an infinite loop in interactive evaluation (as an example).  This API is not for general use and that should be clearly documented.

**Reply by alexrp** [reply]
Relevant context: #41291

**Comment by mangod9** [comment]
> walking the stack to verify the tread is in a safe spot, and raising the exception for thread abort.

What constitutes a "safe" spot? Also would it raise Thread abort or another specific exception for this case?

**Reply by AntonLapounov** [reply]
We have the [`Thread::ReadyForAsyncException`](https://github.com/dotnet/runtime/blob/c8eeba753183d07679b7d2e65cc12d39c948eb4a/src/coreclr/vm/threadsuspend.cpp#L939) function that checks whether a thread is ready for an asynchronous exception. We [prohibit](https://github.com/dotnet/runtime/blob/c8eeba753183d07679b7d2e65cc12d39c948eb4a/src/coreclr/vm/threadsuspend.cpp#L896) even rude thread aborts in some x64 epilogs.

**Reply by AntonLapounov** [reply]
> Also would it raise Thread abort or another specific exception for this case?

Good question. We may consider introducing a new exception class, e.g., `ControlledExecutionException`, instead of reusing `ThreadAbortException`.

**Comment by KevinRansom** [comment]
The proposal looks good.  Although I was wondering why there is Abort and TryAbort, especially since abort is not guaranteed.

I like the idea of a ControlledExecutionException,  obviously ResetAbort should not work for the new exception.  But it should be automatically rethrown if caught and handled.

**Reply by jkotas** [reply]
> But it should be automatically rethrown if caught and handled.

I do not think we should do automatic rethrow if the exception is handled. That was one of the features of thread abort that was very difficult to get right (consider situations like a try/catch block nested inside finally block). I would leave it up to the hosting code to decide whether it wants to inject a new abort exception.

**Reply by AntonLapounov** [reply]
If we allow the exception to be swallowed, `TryAbort` returning `true` would not necessarily mean a successful abort, which might be considered misleading.  Something like `TryInjectException` might be a better naming then. (And if the exception to inject is passed as an argument, we would not need to introduce a special exception type.)

> That was one of the features of thread abort that was very difficult to get right (consider situations like a try/catch block nested inside finally block).

Oh, I have just tried this nested scenario and I think I found a bug.

**Reply by jkotas** [reply]
> Something like TryInjectException might be a better naming then.

Sounds good to me.

**Comment by AntonLapounov** [comment]
> Although I was wondering why there is Abort and TryAbort, especially since abort is not guaranteed.

The underlying [`Thread::UserAbort`](https://github.com/dotnet/runtime/blob/c8eeba753183d07679b7d2e65cc12d39c948eb4a/src/coreclr/vm/threadsuspend.cpp#L1175) function has two arguments: a timeout and a flag indicating whether to do a safe or a rude thread abort.  A safe thread abort may happen  only at a "safe" spot: when the thread is not holding locks and not inside catch/finally regions.  For this reason a safe thread abort is not guaranteed and may timeout.  On the other hand, a rude thread abort ignores those restrictions and it is almost guaranteed to succeed (it is still prohibited to happen in some x64 epilogs).

Do we want to support both types of aborts?  For instance, the debugger first tries to perform a safe thread abort with a half second timeout before initiating a rude thread abort with another half second timeout.

**Reply by jkotas** [reply]
> do a safe or a rude thread abort

The libraries in .NET Framework required a lot of auditing to make "safe" abort actually safe. This hardening does not exist anymore. We got rid of it in number of places since it introduced unnecessary overhead.

> thread is not holding locks 

I believe we got rid of the lock tracking in .NET, is that right? We cannot really do this, unless we bring the lock tracking back. I do not think we want to do that and the lock tracking in .NET Framework was incomplete anyway.

> inside catch/finally regions

The JIT in .NET Core is not fully preserving the information about the try/finally regions. For example, it can fully inline and eliminate finally regions. We would not be able to make guarantees about this.

**Reply by jkotas** [reply]
I think we can have `enum` that allows the caller to suppress (best effort) the exception injection in certain situations, like when the thread is executing inside catch or finally blocks on stack or when leaf method that the thread is executing is in the corelib (again, best effort - it can be imprecise due to inlining).

The app code can then decide in which order and for how long to try the different abort strategies. This is policy heavy piece that we should avoid baking into the runtime.

**Reply by AntonLapounov** [reply]
Thank you for valuable insights.  Since we can no longer guarantee safety for the reasons you described, I am not sure introducing a `enum` for a "slightly safer" strategy is justified.  We might start with providing a rude thread abort only and reconsider if there is good evidence that a better strategy would be helpful.

**Reply by jkotas** [reply]
Is the rude thread abort going to run the exception handling (finally and catch blocks)? (I think it should - just trying to clarify the details.)

**Reply by AntonLapounov** [reply]
If we allow running `catch (Exception) { }` blocks in the code executed by `ControlledExecution.Run()`, we cannot guarantee any progress for the abort since you suggested to not do automatic rethrow.  Can we skip processing EH until we unwind to the `ControlledExecution.Run()` frame?  Or maybe `ControlledExecution.Run()` should use a dedicated thread to run the code and skip EH.  That may still cover the scenario we want to support.

**Reply by jkotas** [reply]
Yes, this again a case of a policy choice between higher risk of corrupted process vs. higher risk of not making progress. 

A related situation is what happens when the thread has interleaving of managed -> unmanaged -> managed on stack? How are we going to unwind that?

**Comment by jkotas** [comment]
> void Abort();
> bool TryAbort(int millisecondsTimeout);

I assume that we would allow `Timeout.Infinite` for the timeout and so `Abort` would be implemented as `TryAbort(Timeout.Infinite)`. I think we can have `TryAbort` only.

**Reply by AntonLapounov** [reply]
If we provide a rude thread abort only, I guess we could keep just `Abort`, because a rude thread abort is quite unlikely to not succeed.  Otherwise, I agree we may have `TryAbort` only.

**Reply by jkotas** [reply]
Rude thread abort can take infinite amount of time in pretty normal situations when the thread is stuck in unmanaged code. I think having the timeout on it is a good idea.

**Reply by AntonLapounov** [reply]
Ah, good point.  I'll leave just `TryAbort` then.

**Comment by TonyValenti** [comment]
I would really like to see a core framework / library for handling out of process function invocation. 

I have a (rather specific) library I've built that I use to do the following:
1.  I pass it a type and a method with parameters. 
2.  It creates a brand new process that instantiates the object and hosts it via JsonRpc. 
3.  I connect to the "remote" object and invoke it. 
4.  If I need to, I can simply terminate the entire hosting process.

I have found this really helpful when invoking wonky COM objects that might hang. I also like it because it keeps unsanitary threads/instances out of my core process and contains them in a disposable process.

**Reply by jkotas** [reply]
There is no reason for this to be in a core framework. It can be standalone nuget package just fine.

**Comment by AntonLapounov** [comment]
It seems we have somewhat conflicting design requirements here:

* We do not want to rethrow the injected exception automatically if it was handled. (As that would be error-prone.)
* That means we may not be able to abort, for example, a long-running loop containing try/catch all.
* That suggests we have to provide a rude thread abort as an option.
* Internally, a rude thread abort destroys the thread; it cannot be handled by managed code.
* That suggests we have to run the specified delegate on a separate thread.

Does that make sense?  If yes, we may start with a rude thread abort on a separate thread and see how that goes. @KevinRansom

**Reply by jkotas** [reply]
> we may start with a rude thread abort on a separate thread and see how that goes

Forcing the scriptlet execution to a separate thread would break scripting of UI frameworks. The UI frameworks are typically single-threaded. @KevinRansom Any thoughts about this?

> That means we may not be able to abort, for example, a long-running loop containing try/catch all.

That may be ok. There are always going to be situations where we are not able to abort (e.g. thread stuck in unmanaged code outside our control). And there are situations where it is more risky to abort. My suggestion above was to allow the user code to control the level of risk using an enum.

**Reply by AntonLapounov** [reply]
A rude thread abort without destroying the thread would be a new abort option that is not currently supported by the runtime.  I see more resemblance to `setjmp`/`longjmp` functionality here than to throwing an exception.  That would skip native frames as well, which is riskier but fixes the issue of the thread being stuck in unmanaged code or waiting.  I think that under any design using rude abort on a UI thread would be dangerous.

**Reply by jkotas** [reply]
I would start with the lowest risk strategy that we can come up with, acknowledging the fact that it may not be able to abort in some situations. If we get feedback later that it is not sufficient, add the higher risk strategies as explicit opt-in.

**Reply by AntonLapounov** [reply]
Since @KevinRansom mentioned above that a caught exception should be rethrown, I would like to hear his opinion on supporting non-rethrown exceptions first.  My concern is that if we may need a rude thread abort eventually, the proposed design should not impede implementing it.

**Reply by KevinRansom** [reply]
@jkotas , the scripting workflow is usually compile some stuff and execute it.  I suppose forcing the script execution on to a separate thread, forces the developer to manage IO such that output from the script and the UI is coordinated, but scripts can create new threads and execute code anyway so the issue already exists.  So I doubt that forcing a new thread for execution is really much of a problem.

**Reply by KevinRansom** [reply]
@AntonLapounov my only concern is that because the abort path is an execution it is conceivable that the script author can include code that catches and swallows the exception thus defeating the whole abort thingy.    The desktop runtime used to have unhandlable execeptions, threadabort was one of them.  I expect that ControlledExecutionException is similar, in that we would implement the handler, and user scripts would not be able to handle the exception.

**Reply by jkotas** [reply]
> I suppose forcing the script execution on to a separate thread, forces the developer to manage IO such that output from the script and the UI is coordinated

Yes, forcing execution on a separate thread (that the API creates for you) makes this API bakes policy into the API. If somebody wants to run the execution, they should be responsible for creating the thread.

I think it is very likely that you won't be able to use the API if it forces execution on a separate thread. If you believe that it should not be a problem, could you please make a change in F# interactive that creates a separate thread for each execution and see how many people get broken by it?

**Reply by AntonLapounov** [reply]
One interesting difference between C# and F# Interactive is preserving the `ExecutionContext`.  C# Interactive uses async calls to run submitted snippets and any changes to `ExecutionContext`, e.g., setting the thread's `CurrentCulture`, are reverted between submissions:
```C#
> Thread.CurrentThread.CurrentCulture = new CultureInfo("de-DE");
> Thread.CurrentThread.CurrentCulture.Name
"en-US"
```
while F# Interactive preserves them:
```F#
> Thread.CurrentThread.CurrentCulture <- CultureInfo.GetCultureInfo("de-DE");;
val it: unit = ()

> Thread.CurrentThread.CurrentCulture.Name;;
val it: string = "de-DE"
```
I agree that running on a separate thread may break some scenarios.  The question here is how much code changes and testing is required to allow rude thread aborts (if we support them) without destroying the thread.

**Reply by N-Olbert** [reply]
Does providing the possibility of a rude thread-abort only offers any significant benefits compared to what is already possible using PInvoke (`pthread_create`/`pthread_cancel` on *NIX, and `CreateThread`/`TerminateThread `on Windows) ?

If not why not providing a separate package with the corresponding functionality instead of expanding the framework?

Im leery in general whether it's "good" to (re-)introduce such an API, especially since its intended use ('development scenarios') is quite narrow and - in my opinion not a primary thing which has to be offered by a **general** runtime like .NET (it could be part of the *SDK* though). If this is going to be introduced into the runtime you can be absolutely sure - no matter how clear you advise against its use in the docs - that this API will eventually being abused as a replacement for 'Thread.Abort' even in production scenarios. Since the .NET runtime seems to be a lot less stable against aborts nowadays, I expect application quality to degrade. In .NET Framework the Framework itself was (more or less) “abort-safe”, and the majority of abort issues were caused by application code, now – even if the application code is indeed “abort-safe”- the framework is not unless [the code changes mentioned by @stephentoub ](https://github.com/dotnet/runtime/discussions/66480#discussioncomment-2334930) are going to be reintroduced.

I also doubt that thread-abortion is the correct level of control to stop the execution of arbitrary user code, since only processes can be aborted (killed) for sure. F. e. a (malicious) script could simply be                
```
new Thread(() =>
                {
                    Thread.CurrentThread.IsBackground = false;
                    while (true) { }
                }).Start();
```
For this script thread-abortion / the new API does not provide any benefits – since the script created its own thread, the abortion of the main scripting thread has no effect at all.

**Reply by AntonLapounov** [reply]
@N-Olbert `TerminateThread` may easily corrupt the native runtime to the point where further process execution is no longer possible (e.g., terminating a thread that is in the middle of a garbage collection). Injecting a managed exception is a much safer mechanism.  Please see #41291 for use cases that may benefit from having this functionality.  I agree on your other points that this API is not bulletproof and may be potentially abused.

**Comment by teo-tsirpanis** [comment]
Who marked the discussion as `api-ready-for-review`? They have to make an issue; discussions with this label will not show up in https://apireview.net/.

**Reply by AntonLapounov** [reply]
Yes, I have already realized that. I will create an issue. Thanks!

**Comment by rickbrew** [comment]
I have an amendment I'd like to propose: Can there be a way to disable this? Possibly with a `DOTNET_DisableControlledExecutionAsyncAbort=1`; some kind of process-wide policy, I mean.

My reason for proposing/requesting this is that I do not want plugins in my app to be able to use something this dangerous, as it could have cascading effects on the stability of the app, ultimately resulting in data loss for the user. In my case, the app is Paint.NET. There are _many_ plugins (some users have 1000+ installed!).

The experience level of plugin developers is wide-ranging. Some are expert software engineers, some are students, while many are just hackers having fun with a copy/paste/modify/run/"neat!" loop. Many of the latter do not know C#, or barely know it, and just push things around it until they get their desired visual result (for an effect plugin).

I absolutely do NOT trust plugin developers to write "good" code. Not just fast code. I mean code that doesn't do bad things, intentionally or not. I say this because I've seen a lot of plugins do very seriously weird or even downright evil stuff -- using reflection to grab things inside of API objects or to modify the Paint.NET UI, hooking major event handlers (AppDomain stuff) to achieve local results but with negative global consequences, etc. While porting Paint.NET to .NET 5 last year I ran into a whole lot of this stuff and had to engineer a lot of creative ways to shim the plugins to workaround and avoid some of the worst behaviors (I go as far as literally rewriting the plugin's IL to detour them around bad methods).

I'm seriously worried that some plugin will use this, and then other plugins will copy-by-example. Or it will get inside of some library / nuget package that becomes popular, with the same result. I want to be able to just block any of that silliness from ever entering my process.

I do believe I'd have a workaround if this does not make it passed API review: I have a highly customized system for loading plugins, using `AssemblyLoadContext`s, and I can just block DLLs with known names and/or versions from loading. (I currently use this to prevent a plugin from using already-loaded DLLs from nuget packages that the app uses, in order to avoid version hell -- a plugin must bundle its own copy of, e.g. `Newtonsoft.Json.dll` or `TerraFX.Interop.Windows.dll`). However, I don't know if other apps with plugin ecosystems will necessarily have things set up to easily accomplish that.

**Reply by AntonLapounov** [reply]
Thank you for your input and for describing your scenario.  Since you control the startup code, I am wondering whether the static `ControlledExecution.Disable` method that permanently disables `TryAbort` for the current process would fit the purpose.

**Reply by jkotas** [reply]
If we wanted to control when this capability is available, I would rather do that via config than via an API. Also, I would consider to have it disabled by default.

**Reply by AntonLapounov** [reply]
> I would consider to have it disabled by default.

As Stephen Toub mentioned in the other issue:
> How are the dangers here different from a library, say, P/Invoking to TerminateThread and similarly leaving things in a corrupted state?

**Reply by jkotas** [reply]
I believe that Stephen's comment is about whether or not to have a mechanism to disable this feature.

I was commenting on the design if we choose to implement a mechanism to disable this feature.

**Reply by teo-tsirpanis** [reply]
Answering to @stephentoub, the dangers are different because this is a sanctioned, readily available and cross-platform API of the BCL, while `TerminateThread` is a platform-specific API that is comparatively much harder to use; defining a P/Invoke or downloading an interop package and figuring out how to get the thread handle is not something that can happen by accident. The proposed API is in `System.Runtime.CompilerServices` which makes it harder to find than for example `System.Threading`, but in my opinion, that's still much easier to find than a P/Invoke.

We are not trying to stop _malicious_ users from corrupting the execution environment, that's infeasible. What we want to do is to prevent _unknowledgeable_ users from viewing an API with a limited use case as a quick replacement of `Thread.Abort`, instead of the correct but laborious practice of passing `CancellationToken`s down the call stack. Our project's code might be correct, but a random NuGet package aborting threads is enough to cause subtle and hard-to-diagnose state corruptions to the rest of the app. That's the reason to guard this API under an opt-in runtime switch.

**Reply by rickbrew** [reply]
> Thank you for your input and for describing your scenario. Since you control the startup code, I am wondering whether the static `ControlledExecution.Disable` method that permanently disables `TryAbort` for the current process would fit the purpose.

It depends on how this is packaged. If it's part of the runtime, then everyone (app and ALC-isolated plugins) are sharing the same instance, and this works fine. If this is a nuget package (which I think is the path being taken?), then plugins are bringing their own copy of it + .deps.json, and each plugin that uses this will have their own copy of the library, potentially different versions of it. `CE.Disable` would mean I'd have to bundle the package with the app even though all I want to do is turn it off, and I'd also have to make sure my ALC dependency resolution code always loads my copy of the library (doable, but would still prefer to not bundle the package in the first place).

For the nuget package case, even without Disable()/config-switch I can still detect the DLL by name and block it from loading, along with an error for the plugin author "This package is not permitted because it's too dangerous". The plugin author could then fork-and-rename, of course, if they're determined. Config switch would work better, although they could fork and rename the config switch name ... again, an arms race and there's only so much you can do. Next level mitigation there is for me to block that plugin entirely in my next app update (which I'm willing to do).

So a config switch is probably the most bang-for-buck mitigation I can see right now. I can't auto-block all bad behavior ... a plugin can always do `*(int*)null` or `DeleteDirectory("%DOCUMENTS%")`, etc. A kill-switch just makes this likely that responsible plugin devs (which is most of them!) will be steered away from it, and that inexperienced devs won't end up with bad copy-paste / cargo-cult coding patterns. Malicious devs will still be able to get around it (fork-and-rename), but I can mitigate that with my plugin blocking/banning mechanism.

**Reply by AntonLapounov** [reply]
I understand this argument, but I would rather educate developers via documentation and trust them than disable by default.  If the obscure namespace is not enough, we might call the method `DangerousAbort` or `StateCorruptingAbort` to make it clear the process state will be corrupted.

**Reply by AntonLapounov** [reply]
@rickbrew This will be implemented in System.Private.CoreLib.dll and the native runtime.

**Reply by rickbrew** [reply]
The developers I'm thinking of won't be reading documentation. Most of them are just hacking away at stuff for fun; they're not necessarily Computer Scientists / Software Engineers. Some of them are even teenagers just copy/paste/hacking their way in the dark (I was there, it was fun, but I also wrote awful code!).

Case-in-point: Despite WinForms' `MenuStrip` and `ToolStrip` being deprecated back in 2006 with the release of .NET 2.0, a very prominent and popular plugin library was released only a few years ago that was using it. It was a big pain point when I migrated to .NET 5.0.

So a config switch is great, I'd love it, but ultimately I can still block the package by name / assembly name, and escalate from there if necessary. This works because I have a robust AssemblyLoadContext infrastructure around plugin loading and isolation. For other apps that don't have that infra, it may be more difficult.

**Reply by rickbrew** [reply]
So consider my proposal a request for allowing the _application_ to establish the policy for assemblies that are loaded into its process space.

Some other random thoughts ... a plugin could indirectly/inadvertently make use of this ~~unsafe~~ dangerous functionality if they pull in a nuget package that uses it. Trimming also makes this harder to mitigate (in another thread, we're figuring out how to make it easy for plugins to link+trim so their DLLs are self-contained and smaller). A config switch would still work for those cases, except for fork-and-rename, but I wouldn't be able to block the DLL by name since no DLL exists.

**Reply by jkotas** [reply]
The point of the configuration switch is to make sure that libraries referenced by the app cannot just use this, without the app author being aware of it. The documentation or scary API names does not address this concern. 

We have number of features that are dangerous or otherwise problematic under a configuration switches. I do not see one more configuration switch as a big deal.

**Reply by rickbrew** [reply]
> @rickbrew This will be implemented in System.Private.CoreLib.dll and the native runtime.

I missed this comment before my reply after it. 

If that's the case then a `Disable()` method is perfectly alright for me.

**Reply by rickbrew** [reply]
(As long as there's no `Enable()` method of course :) )

**Reply by stephentoub** [reply]
> We have number of features that are dangerous or otherwise problematic under a configuration switches.

Can you share examples of APIs considered dangerous that throw unless a switch is set?

If we want to ship an analyer that warns by default if you use anything we consider super dangerous, fine, we've done that with ModuleInitializer (it could look for known pinvokes, too) . But I don't see how this particular API is so much worse than so many other things you can easily do just by copying/pasting answers from [stack overflow](https://stackoverflow.com/questions/4359910/how-to-abort-a-task-like-aborting-a-thread-thread-abort-method/55709530#55709530).

At the end of the day, I don't care super strongly about it, though.

**Reply by jkotas** [reply]
Number of the switches for security and trimmability make APIs throw. For example, `EnableUnsafeUTF7Encoding`, `BuiltInComInteropSupport`, `EnableUnsafeBinaryFormatterSerialization`.

I understand that these switches are not exactly same as what we are talking about here. These switches have often different defaults based on other build properties, and they were often used to fix historical mistakes. However, if you look at it without taking historic context into account, there is very little difference.

**Reply by stephentoub** [reply]
Thanks.

> However, if you look at it without taking historic context into account, there is very little difference.

FWIW, I disagree. These are about, as you say, fixing mistakes, and sunsetting security risks we actively want to prohibit devs from using, with the eventual goal of removing the functionality entirely. Here we're talking about adding such a switch at the exact same time we're introducing the brand-new functionality.

**Comment by rickbrew** [comment]
Also, the environment variable is just an idea. Neither that nor the DLL blocking in the ALC are perfect solutions, I can already think of a few ways for a plugin to get around either.

If it's possible for this setting to be in `.runtimeconfig.json`, that might be the strongest blocking mechanism.

**Comment by AntonLapounov** [comment]
After working on a prototype and test cases, I started leaning towards the original proposal with `TryAbort` and an automatically rethrown exception.  I think that would lead to smaller and safer code churn on the runtime side.  The existing abort machinery in the runtime automatically rethrows the exception until the `TS_AbortRequested` flag on the thread is cleared.  While it is possible to introduce a new abort mode, besides `TA_Safe` and `TA_Rude`, that will clear the flag before the first throw (I think `Thread::HandleThreadAbort` is the right place for that), @KevinRansom's team will have to reimplement rethrowing in their code generator to undo our suppression. On the other hand, if we simply reuse the existing `TA_Safe` abort mode and rethrow the exception until we unwind to the corresponding `ControlledExecution.Run` frame, where we will clear the `TS_AbortRequested` flag, that seems exactly what was requested with no code generator changes required. (We will need to properly handle nested `ControlledExecution.Run` calls though.) And for a general scenario, where catch blocks are not under the host's control, injecting an easily catchable exception seems too weak API.  While the host may attempt to re-inject the exception in a loop, there is no forward progress guarantee.  That makes me think that rethrowing the exception may be a better solution for the general case as well, even if there are some corner cases where rethrowing may not work as expected.  Please let me know if I missed anything. @jkotas

**Reply by jkotas** [reply]
> We will need to properly handle nested ControlledExecution.Run calls though

I do not think we need to. We can say that nested ControlledExecution is not supported.

**Reply by jkotas** [reply]
> if we simply reuse the existing TA_Safe abort mode and rethrow the exception 

Is the TA_Safe abort mode used by the debugger funceval today? 

I see your point that inventing a new scheme is complicated. 

I think it is reasonable to say that this behaves exactly same as funceval abort under debugger. In other words, this is just exposing what's available as debugger API via a managed API. Do you think we can do that?

**Reply by mangod9** [reply]
@janvorli here since some of this TA logic might change as part of CET work, but doesnt look like it will affect this feature.

**Reply by AntonLapounov** [reply]
> Is the TA_Safe abort mode used by the debugger funceval today?

Yes. For some reason GitHub's search does not find that, maybe because the file is too long: https://github.com/dotnet/runtime/blob/77a8d3c818317ac55b2f0c2aca2d68526792a3ab/src/coreclr/debug/ee/debugger.cpp#L15157

Using the same call to abort `ControlledExecution.Run` seems to work and requires minimal changes on the native side.  (Introducing a new thread abort mode is also possible if there is a compelling reason.) As Manish suggested, we may put the `[RequiresPreviewFeatures]` attribute on the `ControlledExecution` class and get feedback on this API before finalizing it.

**Comment by AntonLapounov** [comment]
I have shared my prototype with @KevinRansom the last week.  If anyone is interested to try or review, please take a look: https://github.com/AntonLapounov/runtime/commit/f6025e901239c477f927d78cb05bdea10c27fb4e.  As I mentioned earlier, reusing existing `UserAbort` semantics as much as possible allowed to minimize code churn in the native runtime.

I am wondering if we should consider making public the switch to disable `finally` cloning in the JIT compiler.  Right now, execution of `finally` statements during `TryAbort` depends on whether it is a debug or release build, the length of the `finally` statement, etc.  Disabling the switch would at least give some predictability in that respect (even though we have promised no guarantees).

**Reply by jkotas** [reply]
I would stick to promising no guarantees.
